### ISSUE TITLE
unnecessary_cast: does not handle `std::ptr::null{_mut}`

### ISSUE THREAD

**Issue by KisaragiEffective** [rust-lang/rust-clippy#11912]
### Summary

`unnecessary_cast` does not insert actual type to `std::ptr::null{_mut}`'s turbofish.

### Reproducer

I tried this code:

```rust
#![deny(clippy::unnecessary_cast)]
fn x() {
    let a = std::ptr::null() as *const u8;
    let b = std::ptr::null_mut() as *mut u8;
}
```

I expected to see this happen:

```
error: casting raw pointers to the same type and constness is unnecessary (`*const u8` -> `*const u8`)
 --> src/lib.rs:3:13
  |
3 |     let a = std::ptr::null() as *const u8;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `std::ptr::null::<u8>()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![deny(clippy::unnecessary_cast)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^

error: casting raw pointers to the same type and constness is unnecessary (`*mut u8` -> `*mut u8`)
 --> src/lib.rs:4:13
  |
4 |     let b = std::ptr::null_mut() as *mut u8;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `std::ptr::null_mut::<u8>()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
```

Instead, this happened:

```
error: casting raw pointers to the same type and constness is unnecessary (`*const u8` -> `*const u8`)
 --> src/lib.rs:3:13
  |
3 |     let a = std::ptr::null() as *const u8;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `std::ptr::null()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
note: the lint level is defined here
 --> src/lib.rs:1:9
  |
1 | #![deny(clippy::unnecessary_cast)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^

error: casting raw pointers to the same type and constness is unnecessary (`*mut u8` -> `*mut u8`)
 --> src/lib.rs:4:13
  |
4 |     let b = std::ptr::null_mut() as *mut u8;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `std::ptr::null_mut()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
```

### Version

```text
rustc 1.74.0 (79e9716c9 2023-11-13)
binary: rustc
commit-hash: 79e9716c980570bfd1f666e3b16ac583f0168962
commit-date: 2023-11-13
host: x86_64-unknown-linux-gnu
release: 1.74.0
LLVM version: 17.0.4
```

### Additional Labels

@rustbot labels +I-suggestion-causes-error

**Comment by cocodery** [comment]
Actually, it's more like a generic case more than `std::ptr::null` or `std::ptr::null_mut`.
If in function signature, it return `generic type ptr`, `pointer cast` can be written as `func::<ptr-type>(..)` instead of `func(..) as ptr-type`.
So what need to do is, get the function reture type in `hir` and check whether is a generic type.
But I have no idea about how to get `Item::Fn` to get function information in `Cast` context, is there any suggestions?

**Comment by cocodery** [comment]
It's a generic case more than `std::ptr::null{_mut}`,
but is there any convenient way to check whether `callee` is a generic function?

**Comment by ada4a** [comment]
Interestingly enough, [`ptr_as_ptr`](https://rust-lang.github.io/rust-clippy/master/index.html#/ptr_as_ptr) does give the better suggestion. Running it on the reproducer, I get:
```
warning: casting raw pointers to the same type and constness is unnecessary (`*const u8` -> `*const u8`)
 --> src/lib.rs:5:13
  |
5 |     let a = std::ptr::null() as *const u8;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `std::ptr::null()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast
note: the lint level is defined here
 --> src/lib.rs:2:9
  |
2 | #![warn(clippy::unnecessary_cast, clippy::ptr_as_ptr)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^

warning: `as` casting between raw pointers without changing their constness
 --> src/lib.rs:5:13
  |
5 |     let a = std::ptr::null() as *const u8;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try call directly: `std::ptr::null::<u8>()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_as_ptr
note: the lint level is defined here
 --> src/lib.rs:2:35
  |
2 | #![warn(clippy::unnecessary_cast, clippy::ptr_as_ptr)]
  |                                   ^^^^^^^^^^^^^^^^^^

warning: casting raw pointers to the same type and constness is unnecessary (`*mut u8` -> `*mut u8`)
 --> src/lib.rs:6:13
  |
6 |     let b = std::ptr::null_mut() as *mut u8;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `std::ptr::null_mut()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_cast

warning: `as` casting between raw pointers without changing their constness
 --> src/lib.rs:6:13
  |
6 |     let b = std::ptr::null_mut() as *mut u8;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try call directly: `std::ptr::null_mut::<u8>()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_as_ptr
```
Furthermore, the suggestions of `ptr_as_ptr` are even machine-applicable, so `cargo clippy --fix` does add the `::<u8>`

`ptr_as_ptr` is `pedantic` though, so you won't see its suggestion by default, which is unfortunate

**Comment by KisaragiEffective** [comment]
`ptr_as_ptr` picks the target type from original code and paste it onto the turbofish of the method:
https://github.com/rust-lang/rust-clippy/blob/847ca731dff5fb7ee00621a1b04c8efb203585bc/clippy_lints/src/casts/ptr_as_ptr.rs#L66-L90

but `unnecessary_cast` does not, simply erasing a cast:
https://github.com/rust-lang/rust-clippy/blob/847ca731dff5fb7ee00621a1b04c8efb203585bc/clippy_lints/src/casts/unnecessary_cast.rs#L59

Perhaps we need to do same thing on `unnecessary_cast`.
